### PR TITLE
[dynamo][reland] `ExecutorchCallDelegateHigherOrderVariable` - add sanity check that input and output tensors are disjoint

### LINKED ISSUE #111917: [dynamo] `ExecutorchCallDelegateHigherOrderVariable` may not preserve `FakeTensor` identity
### 🐛 Describe the bug

`ExecutorchCallDelegateHigherOrderVariable` will attempt to retrieve the real values as inputs to the executorch module, and then deep copy fake the result into new fake values. However, it is possible that the `executorch_call_delegate` function might perform in-place ops. 

The new FakeTensors would no longer share the same identity as the input fake tensors, violating our mechanism for Dynamo trace-time aliasing detection. 

No repro as of yet, this is just speculative

Solution: manually dedup by matching the `fake <-> real <-> fake` relationship between inputs and outputs.

### Versions

main

cc @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan

### PR DESCRIPTION
**Pull request by angelayi**
Summary: Reland of https://github.com/pytorch/pytorch/pull/111960, Fixes https://github.com/pytorch/pytorch/issues/111917

Original PR broke some internal tests which the current diff has resolved.

Test Plan: CI

Differential Revision: D51473196

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @jon-chuang 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by jon-chuang — APPROVED** [review]
Thanks for investigating - guess it's fixed on Executorch end.

Needs approval from core maintainer.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**
